### PR TITLE
Refactor conversion function

### PR DESCRIPTION
**Pull request by sgugger**
# What does this PR do?

This PR introduces a refactor of the conversion functions to be able to re-use with safetensors. It comes with zero change of code inside but the main takeaway is to build two functions that take a PyTorch-formatted (resp. TF-formatted) state dict and load it in a TF (resp. PyTorch) model. The state dict is just a dictionary name to tensor, and the tensor car be a NumPy array (as it was) or a torch/tf tensor (which it will be with safetensors).

### CODE REVIEW

File: src/transformers/modeling_tf_pytorch_utils.py [hunk 1]
```diff
@@ -66,10 +67,12 @@ def convert_tf_weight_name_to_pt_weight_name(tf_name, start_prefix_to_remove="",
     if len(tf_name) > 1:
         tf_name = tf_name[1:]  # Remove level zero
 
+    tf_weight_shape = list(tf_weight_shape)
+
     # When should we transpose the weights
-    if tf_name[-1] == "kernel" and tf_weight_shape is not None and tf_weight_shape.rank == 4:
+    if tf_name[-1] == "kernel" and tf_weight_shape is not None and len(tf_weight_shape) == 4:
```
Comment (LysandreJik) [inline src/transformers/modeling_tf_pytorch_utils.py:73]: Why this change? (not saying it's unwarranted, just curious)

**Review by LysandreJik — APPROVED** [review]
That's cool, thanks for handling this refactor separately! Looks good to me, feel free to merge.

File: src/transformers/modeling_tf_pytorch_utils.py (hunk 1 above)
Comment (sgugger) [reply src/transformers/modeling_tf_pytorch_utils.py:73]: Cause the object passed can be a regular shape, not necessarily a TensorFlow shape, and as such, does not always have a rank attribute.
